activerecord/lib/active_record/associations/preloader/through_association.rb

```diff
@@ -87,7 +87,13 @@ def through_scope
             else
               unless reflection_scope.where_clause.empty?
                 scope.includes_values = Array(reflection_scope.values[:includes] || options[:source])
-                scope.where_clause = reflection_scope.where_clause
+                base_wheres = if reflection.klass.finder_needs_type_condition?
+                                reflection.klass.send(:relation_with_type_condition).where_clause
+                              else
+                                ::ActiveRecord::Relation::WhereClause.empty
+                              end
+
+                scope.where_clause = reflection_scope.where_clause - base_wheres
```

Comment:
How about it?

```diff
--- a/activerecord/lib/active_record/associations/preloader/through_association.rb
+++ b/activerecord/lib/active_record/associations/preloader/through_association.rb
@@ -87,6 +87,9 @@ def through_scope
             else
               unless reflection_scope.where_clause.empty?
                 scope.includes_values = Array(reflection_scope.values[:includes] || options[:source])
+                if reflection.klass.finder_needs_type_condition?
+                  reflection_scope = reflection_scope.unscope(where: reflection_scope.klass.inheritance_column)
+                end
                 scope.where_clause = reflection_scope.where_clause
               end
```

`WhereClause#-` not always works properly unlike `WhereClause#except`.